### ISSUE TITLE
Toast TapToDismiss does not override the global config

### ISSUE THREAD

**Issue by CaselIT** [Stabzs/Angular2-Toaster#45]
If a toast defined a different value for `tapToDismiss` in its `toasterConfig` compared to the global configuration, this is ignored.

My use case: I normally do not want to allow tap To Dismiss, but on some toasts I would like to perform an action on click, but is currently not possible since `tabToDismiss` is set to `false`.

Alternatively, execute the `clickHandler` when present even if `tabToDismiss` is set to `false`, since this option by its name should only enable or disable the dismiss on click not the execution of a custom `clickHandler`

Thanks

**Comment by Stabzs** [comment]
Correct, this behavior is currently not implemented.

The `toasterConfig` property attached to a toast instance should be treated as read-only.  It is not exposed as part of the modifiable API since it is only attached AFTER the toast instance has been passed into the respective container.  As a result, you should never attempt to modify configuration via that property.

Now, that said, your request does illustrate an interesting issue that was baked into the original AngularJS-Toaster (I currently have an outstanding issue to fix this there as well), which may have been carried over during the toastr port process.

I think there are a number of potential ways to deal with this, some of which you have already called out.

- Use `BodyOutputType.Component` and attach the click handler to your custom component.  You could define whatever click behavior you would like via this method.  However, this would limit your clickable footprint to the toast body.  This might not be enough to meet your needs.
- Add a toast configuration parameter to the toast instance to allow `tapToDismiss` to be overridden at the per-toast level.
- Change the meaning of `clickHandler` to be `removeToastClickHandler`.  Based on the current semantics, the click handler is intended to be used to evaluate whether or not the toast can in fact be dismissed, should the handler exist.  However, you want to use the handler in a completely different manner, decoupled from the existing meaning.  You simply want a global click handler that you can hook into, which really doesn't exist today.

The third option may be the correct course since it adds more functionality.  This would be a breaking change since the API would need to be modified going forward.

**Comment by CaselIT** [comment]
At the moment the workaround I'm using is to set `tabToDismiss` to true when I want to enable the `clickHandler` on the toast, and set it to the previous value once the toast is no longer shown.

>The toasterConfig property attached to a toast instance should be treated as read-only. It is not exposed as part of the modifiable API since it is only attached AFTER the toast instance has been passed into the respective container. As a result, you should never attempt to modify configuration via that property.

I think that this could be changed be to allow the override of the global config on a per toast base so that the resulting configuration is the join of the two.
I realize that probably this solution is the most demanding in term of effort, but I think that it's the most flexible one.

**Comment by Stabzs** [comment]
I'm glad you have a workaround for the moment, but that's obviously far from optimal and should be addressed.

As for your suggestion, I'm not certain I agree.

A number of the configuration properties are to be used by the container, not the toast itself.  I believe the following list is exhaustive:

```typescript
limit: number;
newestOnTop: boolean;
typeClasses: Object;
iconClasses: Object;
defaultTypeClass: string;
positionClass: string;
animationClass: string;
preventDuplicates: boolean;
toastContainerId: number;
```

Blindly overriding any of these defaults could cause the toast pop instance to not render correctly, disrupt other toast instances, or even completely break that instance of the toast container.

Cherry-picking which attributes to override is also not preferable, since this clutters the API and makes it more confusing for the user to determine how to configure a toast properly.

This becomes even messier when you take into account that a toast could potentially update or overwrite the configuration of multiple toast containers.

I am all for flexibility.  But I also went to great lengths to reduce the footprint of the API and keep it cleaner.  I don't want to trade in explicitness for flexibility.

**Comment by CaselIT** [comment]
I see your point. I had not though about the container configurations options. Considering them I agree with you. An alternative could be to separate them in container configuration and toast configuration. An option to reduce breaking changes could be to ensure that the toast configuration is a subset of the complete container config, so that only the relevant option can be overridden by the toast. But I'm not sure if this could be achieved.

>Change the meaning of clickHandler to be removeToastClickHandler. Based on the current semantics, the click handler is intended to be used to evaluate whether or not the toast can in fact be dismissed, should the handler exist. However, you want to use the handler in a completely different manner, decoupled from the existing meaning. You simply want a global click handler that you can hook into, which really doesn't exist today.

I guess there are two possible options for this one: call the `clickHandler` regardless of the `tapToDismiss` value or create a second handler that does not control the toast closure but is always called.
The first option avoids duplication but it could become problematic to allow clicks on some containers and not on others. The second option avoids this issue, but introduces two handlers that are very similar.

**Comment by Stabzs** [comment]
I apologize for the delayed response.  I've been considering the current alternatives as well as exploring new alternatives.

I think you've summed up the two potential options quite well.

I am concerned that the first option would certainly introduce confusion into the API and cause "buggy" interactions since it would be difficult to use in conjunction with `tapToDismiss` in all scenarios.

On the other hand, like you said, introducing a new click handler that is similar yet different is confusing and unclear.

I think there may actually be an additional option that hasn't been explored yet and may provide a path forward in a backwards-compatible manner.

Currently, the signature for the `clickHandler` is as follows:

`(toast: Toast, isCloseButton?: boolean) : boolean`

A possibility is to add a third parameter that is defaulted to true during toast construction unless explicitly overridden:

`(toast: Toast, isCloseButton?: boolean, isRemovalHandler?: boolean) : boolean`

(name pending, not sure I care for it)

If undefined or null, default to true.  If false, preserve the false value.  

This would allow the function to be backwards' compatible-friendly, preventing a breaking change.  In addition, if a user wanted to simply use it as a normal click handler separate from removal of the toast, they would pass true.

The one major disadvantage to this solution is that there can't be separate handlers for removal and normal clicks.  I'm not sure if this important at all or if there's even a use case for it, but it is a limitation.  Another disadvantage is that the API becomes slightly less clear and explicit.

**Comment by friksa** [comment]
Instead of adding more parameters that might not be used by everyone, it might make sense to use an options parameter instead.  This allows a javascript object that will allow future extensions without api breakage.  Also, it makes it easy to ignore options not used without having to specify many null params.  And the code is more self-documenting.

````
(toast: Toast, isCloseButton?: boolean, options?: any) : boolean

use as:
(toast, isCloseButton, {isRemovalHandler:true})
````

You could also roll isCloseButton into it... and detect if the 2nd param is a boolean or an object.

Just an idea to future proof it.

**Comment by Stabzs** [comment]
@friksa great suggestion.  Perhaps it can be taken one step further.

If the current `isCloseButton` parameter instead becomes a union type (for now), `isCloseButton` can be included into the options object while preserving backwards-compatible behavior.


'(toast: Toast, isCloseButton?: boolean | IClickOptions , options?: IClickOptions) : boolean`

This could be over-complication for the sake of preserving backwards-compatibility and if you feel like it is, then perhaps backing off to removing the `isCloseButton` parameter and coupling it with the options is the right approach.

**Comment by friksa** [comment]
Absolutely.  That is what I tried to say with old javascript terms (including isCloseButton in it and detecting boolean vs object).  Union type is the TypeScript way and I am still learning a lot of new terms :)

**Comment by Stabzs** [comment]
@friksa Actually, I take this back.  It can all be encapsulated into the click options since `isCloseButton` is an internally used parameter; it isn't part of the public API.

This seems like the most reasonable approach.  Thank you for the object options suggestion :)

**Comment by Stabzs** [comment]
@friksa It's a great suggestion and probably the safest way forward for now.  Thank you for the options object suggestion!

**Comment by CaselIT** [comment]
I may be missing something, but is't it the library who calls the `clickHandler`? How can I pass `true` to the new option to it?

And the whole problem is that at the moment the library does not call it when `tapToDismiss` is set to `false`.

Again sorry if what I wrote does not make sense.

**Comment by Stabzs** [comment]
@CaselIT You aren't missing anything at all.

Currently, the library calls the click event which in turn invokes the click handler if it is present.  In order to make something like this work, the toast creation options would need to be expanded.

Currently, when providing a click handler, it is passed in the following manner:

        let toast: Toast = {
            type: 'info',
            clickHandler: (toast , isCloseButton) => {
                console.log("isCloseButton : "+isCloseButton);
                return false;
            }   
        }; 
   
        this.toasterService.pop(toast);

Unfortunately, as you noted, yes, the library invokes the handler itself, deciding what to pass.

I believe that the way in which the proposal would work would be the following:

    let toast: Toast = {
            type: 'info',
            clickOptions: {
                tapToDismiss: true,
                isRemovalHandler: true
                // or some other combination that makes a bit more sense
            },
            clickHandler: (toast , clickOptions?: IClickOptions) => {
                // consume anything off of the click options as you like
                // can be useful for additional scaling.
                console.log("isCloseButton : "+isCloseButton);
                return false;
            }   
        }; 
   
        this.toasterService.pop(toast);

The library would be responsible for preserving the new clickOptions parameter on the toast option.

And when deciding what options to use, the library's click handler would be modified to something like the following:

    click(toast: Toast, clickOptions?:  IClickOptions) {
        if (!clickOptions.isRemovalHandler) {
            toast.clickHandler(toast, clickOptions)
        }
        // else, proceed with removal flow
        // ...
    }

The reason for the need for the union type is that consumers of the library have already been passing in `isCloseButton` for their handlers.  Changing the signature without the union type would break their implementations.  That said, if this is also causing too much complexity, perhaps a breaking change is necessary.

@CaselIT would this solve the problem in a readable, consumable manner?

**Comment by CaselIT** [comment]
@stabzs I think that would work. :+1:
It's just quite similar to the other proposal.

How about a function instead of `clickOptions.alwaysCallClickHandler`, something like `callClickHandler(isCloseButton: boolean): boolean`?

Just food for thought.

**Comment by Stabzs** [comment]
@CasellT I'm not sure I follow. Are you suggesting that instead just provide a callback function on the options that could be hydrated by the user if they choose?

**Comment by CaselIT** [comment]
Yes, that could be an option. The library could provider the default implementino with the same behaviour as now, but an user could change it to adapt it to its needs. In the proposed function above I forgot the toast argument.

A possible implementation could be
```ts
click(toast: Toast, isCloseButton?:  boolean) {
    if (toastConfig.callClickHandler(toast, isCloseButton)) {
        toast.clickHandler(toast, isCloseButton)
    } // else continue with normal remove
}
```

**Comment by Stabzs** [comment]
@CaselIT I think we came full circle a bit too much on this.  

In the above proposal, there's additional complexity in the `callClickHandler` function.  What would it be evaluating?  It seems like this is probably too much complexity and it doesn't accomplish much.

I think either it needs to be a simple flag indicating whether or not to call the click handler or move back to two click handlers.  The function option doesn't seem like it will add much value.

**Comment by CaselIT** [comment]
@Stabzs I agree with you. I thought again about it and it's unnecessary complex the function option.
The other alternatives are both simpler.

**Comment by Stabzs** [comment]
This has been closed in version 11.0.0
